Review large alignment code in GC

AntonLapounov opened the pull request:
Review code related to large alignment.  Improve comments and fix a couple of places where large alignment was not taken into account.  Set `pad_in_front` to zero if `SHORT_PLUGS` is not defined.

AntonLapounov committed:
Fix GC heap corruption on ARM.

The allocate_in_free case in allocate_in_expanded_heap incorrectly calculated the large (double) alignment padding size when limiting the plug size (SHORT_PLUGS) if set_padding_on_saved_p was true:

set_padding_in_expand (old_loc, set_padding_on_saved_p, pinned_plug_entry); // Sets the padding flag on the saved plug
...
pad += switch_alignment_size (is_plug_padded (old_loc)); // Gets the padding flag from the old (different!) plug

That caused access violation during a heap walk since g_gc_pFreeObjectMethodTable was not placed at the right address.

Also clean up related code ensuring large (double) alignment.

AntonLapounov commented:
@Maoni0 Should we also fix `real_size` calculation in `allocate_in_older_generation`?  Namely, add `Align (min_obj_size)` only if `SHORT_PLUGS` is defined and additionally add `switch_alignment_size()` if `RESPECT_LARGE_ALIGNMENT` is defined.  The present code is below:

```C++
int pad_in_front = ((old_loc != 0) && ((from_gen_number+1) != max_generation)) ? USE_PADDING_FRONT : 0;

size_t real_size = size + Align (min_obj_size);
if (pad_in_front)
    real_size += Align (min_obj_size);
```

Maoni0 commented:
oops, I didn't notice you changed so much more than what we talked about (I thought you just had that one line fix). I need to take another look.

AntonLapounov commented:
Submitted the fix for GC heap corruption in a separate PR.

Maoni0 commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -2053,28 +2053,35 @@ ptrdiff_t round_down (ptrdiff_t add, int pitch)
-#else
-    UNREFERENCED_PARAMETER(p1);
-    UNREFERENCED_PARAMETER(p2);
+    return ((((size_t)old_loc ^ (size_t)new_loc) & DATA_ALIGNMENT) == 0) || (old_loc == 0);
```
I would not change & 7 to & DATA_ALIGNMENT because this gets a different answer on 64-bit. I don't think we are defining RESPECT_LARGE_ALIGNMENT on 64-bit but I would still avoid this kind of functional changes.

if you really want, you could add in the comments that by large alignment we really mean 8-byte alignment.

personally I would not do this change at all - now you have to make sure that you pass old_loc as the 1st parameter instead of passing them in either order and just getting an answer about their alignments. but if it really makes you happy I'm ok with it.

also I would do the (old_loc == 0) first -

`return ((old_loc == 0) || ((((size_t)old_loc ^ (size_t)new_loc) & 7) == 0));`

VSadov replied:
Yes, it feels like one of the rare cases where literally `7` seems to fit better, because under RESPECT_LARGE_ALIGNMENT it cannot be anything else. 

Maybe worth adding a compile time assert somwhere that RESPECT_LARGE_ALIGNMENT is used only in 32bit. (or is it only in ARM32 ?)

Maoni0 replied:
yes I agree about the compile time assert (for 32-bit, doesn't need to be for only ARM32)

AntonLapounov replied:
Why don't we replace `return DATA_ALIGNMENT` with `return 4` in `switch_alignment_size` then?  It does not make sense to use symbolic constant in one function and numeric constant in another.  Large alignment means double `DATA_ALIGNMENT`.  It does not matter if `DATA_ALIGNMENT` is 4 or 8, it will still work.

AntonLapounov replied:
> I would not change & 7 to & DATA_ALIGNMENT because this gets a different answer on 64-bit.

And I think that is good.  Large alignment means double alignment, not 64-bit alignment.  If it was intended to mean 64-bit alignment, then `switch_alignment_size` would do bitwise-or with 4, not with `DATA_ALIGNMENT`.

AntonLapounov replied:
> now you have to make sure that you pass old_loc as the 1st parameter instead of passing them in either order

This is a valid argument; the function name does not reflect this argument asymmetry.  I put `old_loc == 0` at the end for performance, because I expect it to be a rare condition.  On the other hand, I expect `old_loc` and `new_loc` to have the same large alignment in 50% cases.

Maoni0 replied:
> Why don't we replace return DATA_ALIGNMENT with return 4 in switch_alignment_size then

I'd much rather you make that change than this one if it really makes you feel better 'cause I don't advise to make carelessly functional changes. large alignment means 4-byte vs 8-byte, it's not tested on any other alignments. while your change gives the right answer for what you think it means, it's still a functional change that I'd not make in a cleanup PR.

btw your change in switch_alignment_size is also a functional change. I don't think it would cause any problems and I think it's a good change. again, this is not something I'd do in a cleanup PR but I'd leave it up to you.

AntonLapounov replied:
> if it really makes you feel better

Sigh. It is not about "feeling better", it is about following good coding guidelines. There are two closely related functions, `same_large_alignment_p` and `switch_alignment_size`, which together ensure double alignment of objects. It makes no sense to use symbolic constant in one of them and numeric constant in another.

>  large alignment means 4-byte vs 8-byte

Feature `RESPECT_LARGE_ALIGNMENT` was introduced with the following comment: "Respect double alignment of objects during compaction", which means alignment to `2*DATA_ALIGNMENT` bytes.  Yes, we are currently using it only when `DATA_ALIGNMENT == 4`.  It does not mean it won't work with minimal changes for, say, `DATA_ALIGNMENT == 8`.

>  it's still a functional change that I'd not make in a cleanup PR

I don't quite understand this argument; in my opinion this sort of functional changes are OK when cleaning up.  I can rename the PR to "Clean up and small change..." if you think that is clearer.

Maoni0 replied:
my concern about combining cleanup with functional changes isn't the commit naming, it is that it changes the amount of risk that the change introduces and thus the amount of testing that the change requires. for a cleanup only change we are asserting with code review alone that the code will produce the same output in all circumstances. for a functional change that is no longer the case. changing the implementation of switch_alignment_size to return a size that's smaller than min obj size could certainly cause problems that are not small (as I said, I don't actually think it would cause any problems but it's the kind of changes that have the potential to cause big problems, unlike the non functional changes). given the high bar for correctness in the GC you would need to do stress testing for this, and a small but not zero chance that the test will produce failures which then you need to be debug. IMO this is a considerable amount of additional work that you would undertake for a very limited benefit. my suggestion is that it would be more efficient to forgo the functional part of the change but it is your call.

AntonLapounov replied:
> changing the implementation of switch_alignment_size to return a size that's smaller than min obj size could certainly cause problems that are not small

Let's do this then and preserve release behavior:
```C++
size_t switch_alignment_size (BOOL already_padded_p)
{
#ifndef RESPECT_LARGE_ALIGNMENT
    assert (!"Should never be called");
#endif // RESPECT_LARGE_ALIGNMENT
   ...existing code...
}
```
I will scale down the changes and resubmit.

VSadov commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -2053,28 +2053,35 @@ ptrdiff_t round_down (ptrdiff_t add, int pitch)
+#endif // RESPECT_LARGE_ALIGNMENT
 }
 
+// Returns the padding size required to respect large (double) alignment.
```
This is a bit confusing. I think this applies not only to doubles. It is more general requirement to align 64bit values on 8 byte locations on 32bit system when on ARM32

AntonLapounov replied:
@VSadov This feature is not about alignment of System.Double objects, it is about of **double alignment** of objects.  It was introduced in 2003, many years before any ARM32 work.

VSadov replied:
I have never seen this used for anything else. 

The description also says
```
// We need the following feature as part of keeping 64-bit types aligned in the GC heap.
#define RESPECT_LARGE_ALIGNMENT //used to keep "double" objects aligned during
                                //relocation
```
It is more clear now why these comments refer to "double". I always thought the comments use `System.Double` as an example.

AntonLapounov replied:
@VSadov I agree this historical comment is confusing now and worth updating.

AntonLapounov replied:
Fixed the comments:
```C++
#define RESPECT_LARGE_ALIGNMENT //Preserve double alignment of objects during relocation
```
and
```C++
// Returns true if two pointers have the same large (double than normal) alignment.
```

VSadov reviewed (approved):
I am ok with the changes. I will let @Maoni0 decide on the `real_size` calculation question.

AntonLapounov commented:
@VSadov Thank you for reviewing this. I am going to address feedback and likely put `(old_loc == 0)` checks back per Maoni's review.

AntonLapounov committed:
Address feedback.

AntonLapounov commented:
Addressed feedback:
- Reverted `old_doc == 0` and `same_large_alignment_p` changes except for using the symbolic constant. (I have a better change in mind.)
- Improved clarity of comments.
- Removed `TODO: this is also not large alignment ready` comment in the `fit` function. I think dotnet/runtime@9235e35 somewhat addressed this issue. We could be a little more clever but it is not likely to worth the trouble.
- Fixed `real_size` calculation in the `allocate_in_older_generation` function.
- Ensured `pad_in_front` is zero if `SHORT_PLUGS` is not defined.
- Changed the title and description of this PR.

Maoni0 commented:
I'm not seeing the functional change in switch_alignment_size reverted. other than that this looks fine to me. 

I don't think at this point we want to spend more time on the large alignment changes unless we need to. I understand your desire to make the changes but we have to prioritize our time so unless there's business needs I would prefer to focus on other areas instead. thanks for your consideration.

AntonLapounov commented:
@Maoni0 Thank you for reviewing this. Could you please explain what you consider a "functional change" here? Is that replacing
```C++
        return (Align (min_obj_size) +((Align (min_obj_size)&DATA_ALIGNMENT)^DATA_ALIGNMENT));
```
with
```C++
        return Align (min_obj_size) | DATA_ALIGNMENT;
```
? Both expressions always produce the same result. The only other diff in `switch_alignment_size` is adding a debug-only assert, which does not affect release behavior.
